blueman/gui/applet/PluginDialog.py

```diff
@@ -284,50 +284,36 @@ def on_toggled(self, _toggle: Gtk.CellRendererToggle, path: str) -> None:
 
         deps = self.applet.Plugins.get_dependencies()[name]
         loaded = self.applet.Plugins.get_loaded()
-        to_unload = []
-        for dep in deps:
-            if dep in loaded:
-                to_unload.append(dep)
+        to_unload = [dep for dep in deps if dep in loaded]
 
         if to_unload:
-            dialog = Gtk.MessageDialog(parent=self, type=Gtk.MessageType.QUESTION, buttons=Gtk.ButtonsType.YES_NO)
-            dialog.props.secondary_use_markup = True
-            dialog.props.icon_name = "blueman"
-            dialog.props.text = _("Dependency issue")
-            dialog.props.secondary_text = \
+            if not self._ask_unload(
                 _("Plugin <b>\"%(0)s\"</b> depends on <b>%(1)s</b>. Unloading <b>%(1)s</b> will also unload <b>"
                   "\"%(0)s\"</b>.\nProceed?") % {"0": ", ".join(to_unload), "1": name}
-
-            resp = dialog.run()
-            if resp != Gtk.ResponseType.YES:
-                dialog.destroy()
+            ):
                 return
+        else:
+            conflicts = self.applet.Plugins.get_conflicts()[name]
+            to_unload = [conf for conf in conflicts if conf in loaded]
 
-            dialog.destroy()
-
-        conflicts = self.applet.Plugins.get_conflicts()[name]
-        to_unload = []
-        for conf in conflicts:
-            if conf in loaded:
-                to_unload.append(conf)
-
-        if to_unload:
-            dialog = Gtk.MessageDialog(parent=self, type=Gtk.MessageType.QUESTION, buttons=Gtk.ButtonsType.YES_NO)
-            dialog.props.secondary_use_markup = True
-            dialog.props.icon_name = "blueman"
-            dialog.props.text = _("Dependency issue")
-            dialog.props.secondary_text = \
+            if to_unload and not self._ask_unload(
                 _("Plugin <b>%(0)s</b> conflicts with <b>%(1)s</b>. Loading <b>%(1)s</b> will unload <b>%(0)s</b>."
                   "\nProceed?") % {"0": ", ".join(to_unload), "1": name}
-
-            resp = dialog.run()
-            if resp != Gtk.ResponseType.YES:
-                dialog.destroy()
+            ):
                 return
 
-            dialog.destroy()
-
-            for p in to_unload:
-                self.applet.Plugins.set_config(p, False)
+        for p in to_unload:
+            self.applet.Plugins.set_config(p, False)
 
         self.applet.Plugins.set_config(name, name not in self.applet.Plugins.get_loaded())
+
+    def _ask_unload(self, text: str) -> bool:
+        dialog = Gtk.MessageDialog(parent=self, type=Gtk.MessageType.QUESTION, buttons=Gtk.ButtonsType.YES_NO)
+        dialog.props.secondary_use_markup = True
+        dialog.props.icon_name = "blueman"
+        dialog.props.text = _("Dependency issue")
+        dialog.props.secondary_text = text
+
+        resp = dialog.run()
```

Comment:
run is a blocking operation. Better to do something like bc25b6d3b74b5c6c7a2d54f731d855db27d4bca8

Reply:
Not sure if I'm missing something, but blocking seems exactly right here as `on_toggled` needs the dialog's response to continue.

Reply:
Of course we could listen to `GtkDialog::response` instead and somehow pass the `to_unload` data to handle the rest of the `on_toggled` code there but what would be the benefit of that?

Reply:
The problem is the whole applet is blocked while the dialog is open. So for example your won't be able to kill it (SIGINT) as that required the main thread to be responding.

Reply:
`killall blueman-applet` while this dialog is open works fine for me, both with this code and in `main` (I did not really change or introduce anything here, just extract and deduplicate the code into a new method).

Reply:
Apperantly I am confused :confused: . Not sure why but even with the blocking call the applet keeps responding.

Reply:
We're inside an event handler at this point. I guess blocking one event handler does not block the event loop, at least in some multi-threaded context, right? :shrug: